### DISCUSSION TITLE
embeddings question

### DISCUSSION THREAD

**Discussion by rtwod2** [pyg-team/pytorch_geometric#3913] [Q&A]
Hi all,

I have some PyTorch geometric graphs, with a varying number of nodes and edges. For nodes, I have 4 features and for edges, I have 1 feature.
I was thinking of creating an embedding layer before my next layer in my architecture, like the code below:

    `    self.embedding_x = nn.Embedding(n_input_dim, hidden_dim)
         self.embedding_e = nn.Embedding(e_input_dim, hidden_dim)
`
The problem is that I can't figure out what n_input_dim and e_input_dim should I have.
I was thinking that I should have for node embeddings for example: [#nodes_in_each_batchgraph, hidden_dim], and because these graphs are of a varying number of nodes I thought that n_input_dim=max_num_nodes*batch_size. 
But this approach outputs the following error:

site-packages/torch/nn/functional.py", line 2044, in embedding
    return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
IndexError: index out of range in self

So I suppose that this is because it tries to find the embeddings in the y-direction (features), not in the x-direction (num of nodes).

Could you please help me with some ideas of what's going on and how should continue?

**Comment by wsad1** [comment]
Is your goal to use `torch.nn.Embedding` to find an embedding for each node_id in the graph independent of the features. Not sure if this would be meaningful, unless your nodes are always in some predefined order.

Or do you want to apply a linear transform on the features before passing the graph to a GNN?

**Reply by rtwod2** [reply]
Hi, 

My goal is to apply a torch.nn.Embedding on the features before passing the graph to a GNN. 

So my thought was to have an embedding for each node capturing the features for that specific node. Because each node has a row of 4 features that make it different than the other nodes in the graph, I'm thinking that I have to have an embedding that will take those features for each node and capture its semantic meanings before passing the graph to my GNN.

**Reply by rusty1s** [reply]
Can you clarify what you mean by 1 and 4 features? I think what you want to do is to apply multiple `torch.nn.Embedding` and sum the embeddings up, similar to our [`AtomEncoder`](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/graphgym/models/encoder.py#L31-L59) for learning on molecular graphs.

**Reply by rtwod2** [reply]
Yes sure. By 1 and 4 features I mean the below:

My graph corresponds to a program.
For each edge in my graph, I have 1 feature, corresponding to its type.
For each node in my graph, I have 4 features, corresponding to its type, block (where the node in my graph is), the InputBytes, and the Device I'm running my whole graph.

So, for my nodes, I would like to have the embeddings of each one of them and then pass it into my GNN model (whatever this is).
The same with my edge embeddings.

**Reply by rusty1s** [reply]
Got it. It looks like writing something similar to `AtomEncoder` (linked above) is applicable for you, assuming all features are of type integer.

**Reply by rtwod2** [reply]
Yes, It worked, thank you for your help. I made something like AtomEncoder for my Nodes and my Edges as well.

I have just one more question to ask you. Sorry if it's too noobie, but I'm trying to learn. Well, imagine you have data.x, data.edge_attr etc.
After I'm creating my embeddings and my GNN model I want to do some regression, so I have to make a global mean. After my global mean pool layer and some reshaping that I made, I have an output, out of shape torch.Size([16, 3, 64]), where 16 is my batch_size, 3 is the batched graphs and 64 is the hidden_dim. 
So, for my regression, I want to have an output of 2 values. For this, I created a linear layer
`self.out_proj = nn.Linear(hidden_dim, 2)`
inside the forward() the call will be:
`output = self.out_proj(out[:,-1:, :])` # out is of shape torch.Size([16, 3, 64])

Which means, give me 2 results, and I want them to be taken from the last batch graph (out of the 3 batched graphs, give me the 3rd one).

However, I can't understand how these 2 values will be eventually the same as the 2 values from my target. The target corresponds to the y of the 3rd batch graph. 

Will my model give me two values until my loss function will be very small?

**Reply by rusty1s** [reply]
I'm not sure I understand, sorry. It looks like
```python
output = self.out_proj(out[:,-1:, :])
```
will map from `[16, 3, 64]` (what exactly is the second dimension?) to `[16, 2]`, which looks absolutely correct to me (you regress two values for each graph).

**Reply by rtwod2** [reply]
Yes, I believe it's correct. I have a data list of batched graphs, which here is of length 3, so I have 3 batched graphs in my data list in every iteration step, and each of these batched graphs is of batch_size=16 (16 graphs). 64 is the hidden_dim. So 3 is my "group" (If I can say it this way) of batched graphs, and I want to have the .y of the 3rd batched graphs (the regression of two values for each graph, [16,2], two values for each of the 16 graphs - in fact the size is [16,1,2]). 
And I can't understand how the [16,1,2] tensor that I'm getting from my predictions=model(batch_out, targets) will eventually be the same as the 2*16 values from my targets. I mean, these values are irrelevant to the targets. The only thing I can suspect is that we say give me those batch_out values [16,1,2] and try to make them equal with those [16,1,2] targets that I already have. 
Again I'm sorry, cause this is a noobie question, but I'm trying to understand it all the way down.

**Reply by rusty1s** [reply]
Thanks for your response. So, as far as I understand, the graph structure is shared across the batch_size (`16`), and each individual graph is described as a supergraph holding 3 individual graphs (batched graphs). If you are simply using the embeddings of the last graph, you are actually ignoring the embeddings of the first two ones. Is that intended?

Anyway, how the model output and target relate to each other depends on how you encode them, and it looks like I'm missing out on some information on how you do mini-batching to answer this question appropriately. In general, the order of data will not be mixed up, and you can safely assume that `data.x[i]` and `data.y[i]` points to the same example.

**Reply by rtwod2** [reply]
Thank you for your time and your help. 

Well, actually, from my data loader I have a list of batched graphs for every timestamp. So, for every timestamp, I have a list of length 3, where 3 is how many batched graphs I have.
Now, these batched graphs are holding 16 individual graphs (the batch_size). So lists contain 3 batched graphs of 16 graphs each (and in the last step for each timestamp, I have 3 batched graphs of 13 - because of the total # graphs). 
So it's like a sliding window of graphs.
How do you mean that I'm using the embeddings of the last graph? My intention is to gain all the information from all the graphs and then give the two output values for the 3rd one (only 2 values), or because of my encoding the 3rd batched graphs (so 16*2).

**Reply by rusty1s** [reply]
OK, thanks for clarifying. Are your batched graphs inter-connected with each other? Otherwise `out[:, -1, :]` will discard any information of the first two graphs. Otherwise, everything looks fine to me.

**Reply by rtwod2** [reply]
> Are your batched graphs inter-connected with each other? 

Do you mean how the problem is defined, independent of using PyTorch geometric? No, each graph is independent. There is a possibility some of them to be a part of a bigger program, that's why I am thinking of having this sliding window. Otherwise, it would be a regression per graph. So we would have `self.linear3 = Linear(int(dense_neurons/2), 2) ` and  `x = self.linear3(x)` in the end.

Do you think that if my graphs (representing kernels) are independent, would be preferable to have a regression per graph?

**Reply by rusty1s** [reply]
If your graphs are independent, you will need some mechanism in your model to aggregate results from all three graphs, right? Otherwise, the two other graphs are not really being used for making a prediction. For example, I'm thinking of doing something like `out.mean(dim=1)` rather than `out[:, -1]`.

**Reply by rtwod2** [reply]
Hi again. Sorry for my late answer. Yes, you're right, I need somehow to aggregate results from all three batched graphs in my list. I changed the `output = self.out_proj(out[:,-1:, :])` into `output = self.out_proj(out.mean(dim=1))`.
The problem is that I have an enormous loss value, in both cases. 
```python
epoch_loss = 0
for step, (data_list, targets) in enumerate(trainloader):
       optimizer.zero_grad()
       batch_out = model(data_list)
       loss = model.loss(batch_out, targets)
       loss.backward()
       optimizer.step()
      epoch_loss += loss.detach().item()
epoch_loss /= (step+1)
```
where loss is:
```python
 def loss(self, scores, targets):
      loss = nn.L1Loss()(scores, targets)
      return loss
```

The loss is like: Epoch: 0199 loss_train: 5306.8379 time: 0.2092s

I also tested if this was a problem in the simple regression task of having just the data of one graph and predicting the data.y of this graph. 
```python
for step, batch_graphs in enumerate(trainloader):
     ...
     loss_train = model.loss(batch_scores, batch_graphs.y)
     ...
     epoch_loss += loss_train.detach().item()
epoch_loss /= (step+1)
```
But the problem seems to be the same, enormous loss (Epoch: 0199 loss_train: 5101.0034 time: 0.1448s).

Would this be a problem of how my node features are in the graph? ex. I have the information of which device is used for this kernel/graph as a feature in all of the nodes in this kernel/graph, and the same is for the InputBytes. Could this make the training harder?
What should I do to fix the loss problem?

**Reply by rusty1s** [reply]
If your loss is so high, that likely means that your target labels aren't normalized, right? You may want to try that.

**Reply by rtwod2** [reply]
Hi Matthias, Thank you for your answer, you're help is very important for me and I'm really grateful.

Below, I'm referring now to the standard regression of having the data of one graph and predicting the data.y of this graph (not the one with the data list of batched graphs, but it's the same I suppose plus the wrapper).

Yes, you're right, my target labels aren't normalized. I haven't normalized also my other data in pyg graph (node features and edge features). When I normalize only data.y then the loss is getting indeed low. I have two questions about this.

Should I normalize also my node features and edge attributes? I have 4 node features with integers values. I have one of them with very large values, like ex. 4004 or 40004. The other 3 node features are like 0, 2 (small values).

The second question is about my predictions, I'm pasting some of the results of the training part.

shape of batch scores is torch.Size([16, 2]), tensor([[-1.4368e-03,  1.0150e+00],
        [-3.7005e-04,  1.0378e+00],
        [-5.2741e-05,  1.0820e+00],
        [-1.2655e-04,  9.2121e-01],
        [ 1.5668e-03,  8.6904e-01],
        [-3.1354e-04,  1.0360e+00],
        [ 1.1428e-03,  9.4795e-01],
        [ 1.5679e-03,  9.5343e-01],
        [ 5.4678e-05,  1.1758e+00],
        [ 2.6163e-03,  1.1436e+00],
        [ 1.5333e-03,  1.0573e+00],
        [-1.3108e-03,  8.9832e-01],
        [-1.0748e-03,  9.9261e-01],
        [ 7.7956e-04,  8.3808e-01],
        [ 2.8418e-04,  1.0598e+00],
        [-7.8584e-04,  8.6734e-01]], grad_fn=<AddmmBackward0>)
batch_graphs.y is torch.Size([16, 2]), tensor([[1.5458e-05, 1.0000e+00],
        [1.4853e-05, 1.0000e+00],
        [1.4100e-05, 1.0000e+00],
        [1.3513e-05, 1.0000e+00],
        [2.2883e-04, 1.0000e+00],
        [1.1715e-04, 1.0000e+00],
        [7.7977e-05, 1.0000e+00],
        [5.9130e-05, 1.0000e+00],
        [4.8300e-05, 1.0000e+00],
        [4.0377e-05, 1.0000e+00],
        [3.5072e-05, 1.0000e+00],
        [3.1299e-05, 1.0000e+00],
        [2.8044e-05, 1.0000e+00],
        [2.5368e-05, 1.0000e+00],
        [2.3192e-05, 1.0000e+00],
        [2.1451e-05, 1.0000e+00]])
loss_train is 0.04168979823589325
As you can see I have a very small loss value, but I can't see a match between the predicted scores and the data.y.
The target values are time, bytes with this order, so the '1.0000e+00' value for example is a normalized byte number, which in this batch seems to be the same.
Do you have any intuition about this?

**Reply by rusty1s** [reply]
It always depends on the distribution whether one needs to normalize node features, but IMO there isn't very often a reason to not do it. You can also treat it as a hyper-parameter which you can tune. I think your predictions look pretty close already, e.g., `9.5343e-01` is pretty close to `1` already.

**Reply by rtwod2** [reply]
Yes, I got it. I wrote something, so as to be like a hyper-parameter. The thing is that if I'm doing this, and for example use a Normalization or a Standardization, I have float values, but this is a conflict with my Embeddings which I created writing something like [AtomEncoder](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/graphgym/models/encoder.py#L31-L59). 
I was thinking of just normalizing only one out of the four node features, which is the one with very big integer values, but if I'm doing this then I couldn't put it inside the Embeddings, which for now is something that I want.
So, maybe I have to think something else for node feature scaling.

**Reply by rusty1s** [reply]
If you rely on embeddings, there is no need for any normalization.

**Reply by rtwod2** [reply]
Yes, this is my thinking too. I did made some runs, and it works better now. I have some overfitting, but I suppose is because I don't have many graphs for now. But I'll find a solution for this hopefully soon. 
![acc_train_valid](https://user-images.githubusercontent.com/98180692/154868425-c9cfcd14-fbd8-4e94-a074-c3645687a8c7.png)
![train_valid_loss](https://user-images.githubusercontent.com/98180692/154868436-c150f8f0-94eb-4580-a990-9f67b792e454.png)

I read that for PyG >= 1.6 it support sparse-matrix multiplication GNNs. But as a general rule of thumb, this holds true for GNNs that do not make use of the central node features x_i or multi-dimensional edge features when computing messages. [https://pytorch-geometric.readthedocs.io/en/latest/notes/sparse_tensor.html](url)
I'm thinking of making my transformer sparse and instead of computing all the dot products, to compute only the ones which are connected with edges and of course taking into account the edge features. So, I wanted to ask you what do you mean as multi-dimensional edge features, would you give an example of this?

Thank you for being so patient with my never ending questions!

**Reply by rusty1s** [reply]
I mean any case in which you have a rich set of features attached to the edges (rather than just a scalar that denotes the weight of an edge). In the attention-case, it should be still possible to apply sparse-matrix multiplication (since you all you do is learning the respective weight).

**Reply by rtwod2** [reply]
Okayy I got it. So indeed I have my edge features, which are currently two, but my purpose is to have one of them inside the attention, and make changes in your graph transformer you provide. I suppose it's okay. 
I' ll update this discussion hopefully this week. Thank you for your time, and your guidance.
